### PR TITLE
Create new adapter for Sublime

### PR DESCRIPTION
**Pull request by SublimePierreAlexis**
Creation of the new Sublime Adapter

## How to test:

```sh
# In your env
git clone git@github.com:SublimeSkinz/Prebid.js.git

# Install dependencies
npm install

# Everything is already configured in the project, you only have to run tests like this
npm run test
```

[Here is the documentation](http://prebid.org/dev-docs/bidder-adaptor.html#bidder-adaptor-Building-the-Request) to create a new adapter for prebid.js

[Here are some guidelines](https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md) to follow, don't forget to configure your linter and IDE to respect prebid coding style

To check coding style type 
```
gulp lint
```

To check our unit tests with Karma :
```
gulp test
```

To check coverage :
```
gulp test-coverage
```

It will create an index.html file in buid/coverage/karma_html. Then you can open it and navigate to module and check the coverage of our SublimeAdapter.

# What to expect

Linter should be all green.
All tests should be green.
Coverage should be higher than 80 % (mandatory to be merged).

# Going further 

If you want to have an end to end test of our adapter. You can check this pull request 
https://github.com/SublimeSkinz/Emilbus/pull/527

### CODE REVIEW

**Review by SublimeVincent — APPROVED** [review]

File: modules/sublimeBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      height: 1000,
+      creativeId: 1,
+      dealId: 1,
+      currency: serverResponse.body.currency || 'USD',
+      netRevenue: true,
+      ttl: 600,
+      referrer: '',
+      ad: serverResponse.body.ad || '',
+    };
+    if (bidResponse.cpm) {
```
Comment (sublimefengyu) [inline modules/sublimeBidAdapter.js]: Why we don't check this before creating bidResponse ? We can use a condition like `serverResponse.body && serverResponse.body.cpm`

File: modules/sublimeBidAdapter.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+Connects to Sublime for bids.
+Sublime bid adapter supports Skinz and M-Skinz formats.
+
+# Test Parameters
+
+```
+var adUnits = [{
+    code: 'sublime',
+    sizes: [[1800, 1000]],
```
Comment (SublimePierreAlexis) [inline modules/sublimeBidAdapter.md:44]: What about the `640x300` format ? We don't care ?

File: test/spec/modules/sublimeBidAdapter_spec.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        zoneId: 14312,
+        bidHost: 'pbjs.ayads.co.local',
+        callbackName: 'myCallback'
+      }
+    }];
+
+    let request = spec.buildRequests(bidRequests);
+
+    it('should have a get method', () => {
+      expect(request.method).to.equal('GET');
```
Comment (SublimePierreAlexis) [inline test/spec/modules/sublimeBidAdapter_spec.js]: Already tested no ?

File: test/spec/modules/sublimeBidAdapter_spec.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+      requestId: 'xyz654',
+      params: {
+        zoneId: 1
+      }
+    }];
+
+    let request = spec.buildRequests(bidRequests);
+
+    it('should have a get method', () => {
+      expect(request.method).to.equal('GET');
```
Comment (SublimePierreAlexis) [inline test/spec/modules/sublimeBidAdapter_spec.js]: Already tested no ?

**Review by SublimeTheo — APPROVED** [review]

**Comment by SublimeJeremy** [comment]
Add size to AdUnit

File: modules/sublimeBidAdapter.js [hunk 5]
```diff
@@ -9,6 +9,28 @@ const DEFAULT_PROTOCOL = 'https';
 const SUBLIME_VERSION = '0.1';
 let SUBLIME_ZONE = null;
 
+/**
+ * Send a pixel to antenna
+ * @param {String} name The pixel name
+ * @param {viaSublime} viaSublime
+ */
+function sendAntennaPixel(name, viaSublime) {
+  if (viaSublime) {
```
Comment (SublimeJeremy) [inline modules/sublimeBidAdapter.js]: merge lines and send with `triggerPixel` in `else`

File: modules/sublimeBidAdapter.js [hunk 6]
```diff
@@ -9,6 +9,28 @@ const DEFAULT_PROTOCOL = 'https';
 const SUBLIME_VERSION = '0.1';
 let SUBLIME_ZONE = null;
 
+/**
+ * Send a pixel to antenna
+ * @param {String} name The pixel name
+ * @param {viaSublime} viaSublime
+ */
+function sendAntennaPixel(name, viaSublime) {
```
Comment (SublimeJeremy) [inline modules/sublimeBidAdapter.js]: `viaSublime` not necessary, it really depends on `(typeof top.sublime !== 'undefined')`

File: modules/sublimeBidAdapter.js [hunk 7]
```diff
@@ -140,28 +150,23 @@ export const spec = {
...
-        top.sublime.analytics.fire('dirnocpm', {
-          qs: {
-            z: SUBLIME_ZONE
-          }
-        });
+        sendAntennaPixel('bid');
+        bidResponses.push(bidResponse);
       }
     } else {
       // debug pixel no request
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: // debug pixel no response

Comment (SublimePierreAlexis) [reply modules/sublimeBidAdapter.js]: I didn't renamed the pixel, so I let `request` for now

File: modules/sublimeBidAdapter.js [hunk 8]
```diff
@@ -155,25 +160,25 @@ export const spec = {
       };
 
       if (!response.cpm) {
-        sendAntennaPixel('dirnocpm');
+        sendAntennaPixel('dirnocpm', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: requestId is declared in buildRequests method : maybe we have to use bidResponse.requestId here

File: modules/sublimeBidAdapter.js [hunk 9]
```diff
@@ -155,25 +160,25 @@ export const spec = {
...
       }
 
       if (response.timeout) {
         // Debug timeout from the long polling server
-        sendAntennaPixel('dlptimeout');
+        sendAntennaPixel('dlptimeout', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: bidResponse.requestId instead of requestId ?

File: modules/sublimeBidAdapter.js [hunk 10]
```diff
@@ -155,25 +160,25 @@ export const spec = {
...
       } else if (bidResponse.ad.match(regexNoAd)) {
         // Debug LP response no ad (a=0 in the notify)
-        sendAntennaPixel('dlpnoad');
+        sendAntennaPixel('dlpnoad', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: bidResponse.requestId instead of requestId ?

File: modules/sublimeBidAdapter.js [hunk 11]
```diff
@@ -155,25 +160,25 @@ export const spec = {
...
       } else if (bidResponse.ad === '') {
         // Debug no ad in the interpret response, what happenned ?
-        sendAntennaPixel('drespnoad');
+        sendAntennaPixel('drespnoad', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: bidResponse.requestId instead of requestId ?

File: modules/sublimeBidAdapter.js [hunk 12]
```diff
@@ -155,25 +160,25 @@ export const spec = {
...
       } else {
-        sendAntennaPixel('bid');
+        sendAntennaPixel('bid', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: bidResponse.requestId instead of requestId ?

File: modules/sublimeBidAdapter.js [hunk 13]
```diff
@@ -155,25 +160,25 @@ export const spec = {
...
         bidResponses.push(bidResponse);
       }
     } else {
       // debug pixel no request
-      sendAntennaPixel('dirnorq');
+      sendAntennaPixel('dirnorq', requestId);
```
Comment (SublimeTheo) [inline modules/sublimeBidAdapter.js]: if no response (= serverResponse.body) we don't have any requestId so we can't send it

File: modules/sublimeBidAdapter.js [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+      qs: {
+        z: SUBLIME_ZONE
+      }
+    };
+    if (requestId) {
+      params.qs.reqId = encodeURIComponent(requestId);
+    }
+    top.sublime.analytics.fire(name, param);
+  } else {
+    var et = Math.round(sublime.window.performance.now());
```
Comment (SublimePierreAlexis) [inline modules/sublimeBidAdapter.js]: Variable `sublime ` is not defined here, that's why we call manually the antenna pixel via the url

Comment (SublimeTheo) [reply modules/sublimeBidAdapter.js]: Maybe just window.performance.now() will work ?

**Review by SublimeVincent — APPROVED** [review]

**Comment by fabio998** [comment]
Looking for Sublime prebid adapter, I found this fork!
Good stuff guys. Any booked official release date?

**Pull request closed without merging**